Towny/src/main/java/com/palmergames/bukkit/towny/utils/SpawnUtil.java

```diff
@@ -440,19 +445,88 @@ else if (town != null && town.hasSpawn())
 							return player.getWorld().getSpawnLocation();
 					});
 				} else if (town != null && town.hasSpawn())
-					yield CompletableFuture.completedFuture(town.getSpawnOrNull());
+					yield CompletableFuture.completedFuture(getSafeLocation(town.getSpawnOrNull()));
 				else
 					yield CompletableFuture.completedFuture(player.getWorld().getSpawnLocation());
 			case TOWN:
 				if (outpost)
-					yield CompletableFuture.completedFuture(getOutpostSpawnLocation(town, split));
+					yield CompletableFuture.completedFuture(getSafeLocation(getOutpostSpawnLocation(town, split)));
 				else
-					yield CompletableFuture.completedFuture(town.getSpawn());
+					yield CompletableFuture.completedFuture(getSafeLocation(town.getSpawn()));
 			case NATION:
-				yield CompletableFuture.completedFuture(nation.getSpawn());
+				yield CompletableFuture.completedFuture(getSafeLocation(nation.getSpawn()));
 		};
 	}
 
+	/**
+	 * Tries to find a safe location nearby to teleport the player to
+	 * if safety teleport is enabled, otherwise does nothing.
+	 * 
+	 * @param location Starting location
+	 * @return A safe location nearby, same location if already safe
+	 */
+	private static Location getSafeLocation(Location location) {
+		//if safety teleport isn't enabled do nothing
+		if (!TownySettings.getSafetyTeleport()) {
+			return location;
+		}
+		
+		if (isSafeLocation(location)) {
+			return location;
+		}
+
+		Location up = location.clone();
+		Location down = location.clone();
+		
+		// look for 20 blocks up and down for a safe location, 
+		// if you can't find it fail the teleport
+		// maybe add a translation key and add the player 
+		// as a parameter to print him an error message
+		for(int i = 0; i < 20; i++) {
+			up = up.add(0,1,0);
+			down = down.subtract(0, 1, 0);
+			
+			if (isSafeLocation(up)) {
+				return up;
+			}
+			
+			if (isSafeLocation(down)) {
+				return down;
+			}
+		}
+		
+		return null;
+	}
+
+	private static boolean isSafeLocation(Location location) {

```

Comment:
I am a bit apprehensive about this method for a couple reasons:

I'm fairly sure that calling world.getBlockAt(Location) has to load the chunks to learn about the blocks. Part of using PaperLib to teleport players was making teleports less painful on servers. This new system will probably result in losing that performance gain. It is possible to load the WorldCoord that contains the spawn point though so this is something that can probably get solved for.

The other thing I don't like about this system is that it currently tests blocks over and over again.

You give it a block and it tests that block, then the one below it and then the one above it. If none are good blocks, you switch to testing the block above, so block Y+1 is now tested (which it has already been tested from the first isSafeLocation() call. It will then test the block below (which was also already tested the previous run,) and then finally tests the block above. 

There's a lot of repeat testing going on. And it is happening every time someone spawns to the town/nation/outpost.

You could probably save time by passing in a BlockFace.UP or BlockFace.DOWN when you're checking up or down, and then ignore tests based on the direction you're testing in.

Reply:
Yeah that is true, fixing it in a bit